nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateXml.java

```diff
@@ -134,32 +157,73 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
         }
 
         final Schema schema = schemaRef.get();
-        final Validator validator = schema.newValidator();
+        final Validator validator = schema == null ? null : schema.newValidator();
         final ComponentLog logger = getLogger();
+        final boolean attributeContainsXML = context.getProperty(XML_SOURCE_ATTRIBUTE).isSet();
 
         for (FlowFile flowFile : flowFiles) {
             final AtomicBoolean valid = new AtomicBoolean(true);
-            final AtomicReference<Exception> exception = new AtomicReference<Exception>(null);
-
-            session.read(flowFile, new InputStreamCallback() {
-                @Override
-                public void process(final InputStream in) throws IOException {
-                    try {
-                        validator.validate(new StreamSource(in));
-                    } catch (final IllegalArgumentException | SAXException e) {
-                        valid.set(false);
-                        exception.set(e);
+            final AtomicReference<Exception> exception = new AtomicReference<>(null);
+            SafeXMLConfiguration safeXMLConfiguration = new SafeXMLConfiguration();
+            safeXMLConfiguration.setValidating(false);
+
+            try {
+                DocumentBuilder docBuilder = safeXMLConfiguration.createDocumentBuilder();
+
+                if (attributeContainsXML) {
+                    // If XML source attribute is set, validate attribute value
+                    String xml = flowFile.getAttribute(context.getProperty(XML_SOURCE_ATTRIBUTE).evaluateAttributeExpressions().getValue());
+                    ByteArrayInputStream bais = new ByteArrayInputStream(xml.getBytes(StandardCharsets.UTF_8));
+
+                    if (validator != null) {
+                        // If schema is provided, validator will be non-null
+                        validator.validate(new StreamSource(bais));
+                    } else {
+                        // Only verify that the XML is well-formed; no schema check
+                        docBuilder.parse(bais);
                     }
```

Comment:
Since this logic is essentially the same as the logic in the callback in the `else` block, it should be possible to create a shared function that takes an `InputStream`, which could then be called as needed.

Reply:
Implemented single method for shared functionality.